File: app/src/main/java/net/bloople/stories/OutlineAdapter.java
package net.bloople.stories;

import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import org.commonmark.node.Node;

import java.util.ArrayList;
import java.util.List;

import io.noties.markwon.Markwon;

class OutlineAdapter extends RecyclerView.Adapter<OutlineAdapter.ViewHolder> {
    private Markwon markwon;
    private List<Node> nodes;
    private List<Integer> nodeIndexes;

    class ViewHolder extends RecyclerView.ViewHolder {
        TextView textView;
        ViewHolder(View view) {
            super(view);
            textView = (TextView)view.findViewById(R.id.text_view);

            textView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    ReadingStoryActivity activity = (ReadingStoryActivity)view.getContext();

                    activity.scrollToPosition(nodeIndexes.get(getAdapterPosition()));
                    activity.closeDrawers();
                }
            });
        }
    }

    OutlineAdapter(Markwon markwon) {
        this.markwon = markwon;
        nodes = new ArrayList<>();
        nodeIndexes = new ArrayList<>();
    }

    void addAll(List<Node> newNodes, List<Integer> newNodeIndexes) {
        nodes.addAll(newNodes);
        nodeIndexes.addAll(newNodeIndexes);
        notifyItemRangeInserted(nodes.size() - 1, newNodes.size());
    }

    // Create new views (invoked by the layout manager)
    @Override
    public OutlineAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.outline_node_view, parent,
                false);

        return new ViewHolder(view);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        TextView tv = holder.textView;

        markwon.setParsedMarkdown(tv, markwon.render(nodes.get(position)));

        tv.setPadding(
                tv.getPaddingLeft(),
                (position == 0 ? tv.getPaddingBottom() : 0),
                tv.getPaddingRight(),
                tv.getPaddingBottom()
        );
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return nodes.size();
    }
}

File: app/src/main/java/net/bloople/stories/OutlineAdapter.kt
package net.bloople.stories

import io.noties.markwon.Markwon
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import org.commonmark.node.Node
import java.util.ArrayList

internal class OutlineAdapter(private val markwon: Markwon) : RecyclerView.Adapter<OutlineAdapter.ViewHolder>() {
    private val nodes: MutableList<Node>
    private val nodeIndexes: MutableList<Int>

    internal inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var textView: TextView

        init {
            textView = view.findViewById(R.id.text_view)
            textView.setOnClickListener { view ->
                val activity = view.context as ReadingStoryActivity
                activity.scrollToPosition(nodeIndexes[bindingAdapterPosition])
                activity.closeDrawers()
            }
        }
    }

    fun addAll(newNodes: List<Node>, newNodeIndexes: List<Int>) {
        nodes.addAll(newNodes)
        nodeIndexes.addAll(newNodeIndexes)
        notifyItemRangeInserted(nodes.size - 1, newNodes.size)
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(
            R.layout.outline_node_view, parent,
            false
        )
        return ViewHolder(view)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val tv = holder.textView
        markwon.setParsedMarkdown(tv, markwon.render(nodes[position]))
        tv.setPadding(
            tv.paddingLeft,
            if (position == 0) tv.paddingBottom else 0,
            tv.paddingRight,
            tv.paddingBottom
        )
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return nodes.size
    }

    init {
        nodes = ArrayList()
        nodeIndexes = ArrayList()
    }
}